src/main/java/ch/njol/skript/util/PersistentDataUtils.java

```diff
@@ -0,0 +1,605 @@
+					} else if (actualHolder instanceof TileState) {
+						((TileState) actualHolder).update();
+					}
+
+					return true;
+				}
+			} else { // The map is empty, so we should check Metadata
+				Metadatable mHolder = (Metadatable) holder;
+
+				List<MetadataValue> mValues = mHolder.getMetadata(keyName);
+
+				if (!mValues.isEmpty()) {
+
+					Map<String, Object> mMap = null;
+					for (MetadataValue mv : mValues) { // Get the latest value set by Skript
+						if (mv.getOwningPlugin() == Skript.getInstance()) {
+							mMap = (Map<String, Object>) mv.value();
+							break;
+						}
+					}
+					if (mMap == null)
+						mMap = new HashMap<>();
+
+					if (index.equals("*")) { // Remove ALL values
+						if (!mHolder.hasMetadata(keyName))
+							return false;
+						mHolder.removeMetadata(keyName, Skript.getInstance());
+						return true;
+					} else { // Remove value and set again
+						if (!mMap.containsKey(index))
+							return false;
+						mMap.remove(index);
+						if (mMap.isEmpty()) { // No point in storing an empty map. The last value was removed.
+							mHolder.removeMetadata(keyName, Skript.getInstance());
+						} else {
+							mHolder.setMetadata(keyName, new FixedMetadataValue(Skript.getInstance(), mMap));
+						}
+
+						return true;
+					}
+
+				}
+			}
+		} else {
+			NamespacedKey key = getNamespacedKey(name);
+			if (key == null)
+				return false;
+
+			if (actualHolder.getPersistentDataContainer().has(key, SINGLE_VARIABLE_TYPE)) {
+				actualHolder.getPersistentDataContainer().remove(key);
+
+				// This is to update the data on the ItemType or TileState
+				if (holder instanceof ItemType) {
+					((ItemType) holder).setItemMeta((ItemMeta) actualHolder);
+				} else if (actualHolder instanceof TileState) {
+					((TileState) actualHolder).update();
+				}
+
+				return true;
+			}
+
+			// Try to remove Metadata instead
+			if (holder instanceof Metadatable) {
+				Metadatable mHolder = (Metadatable) holder;
+				if (mHolder.hasMetadata(name)) {
+					mHolder.removeMetadata(name, Skript.getInstance());
+					return true;
+				}
+			}
+
+		}
+
+		return false;
+	}
+
+	/**
+	 * Whether or not the holder has a value under the key.
+	 * <b>Note:</b> If the holder doesn't have this value in its PersistentDataContainer, then the holder's metadata will
+	 * instead be checked. If a value is found under metadata, then this method will return true.
+	 * @param holder A {@linkplain PersistentDataHolder}, a {@linkplain Block}, or an {@linkplain ItemType}.
+	 * @param name The key name in the PersistentDataContainer. It will go through conversion to replace some characters.
+	 * @return True if the user has the key and false if they do not (or if a problem occurred e.g invalid holder, name).
+	 */
+	public static boolean has(Object holder, String name) {
+		// Run the get method because of how much needs to be checked/done (avoid code repetition)
+		for (Object object : get(holder, name)) {
```

Comment:
it would probably be more efficient and more clear to have a method that gets the key name for an index, and then just use `PersistentDataContainer#has` so it doesn't have to deserialize just to check for existence